crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,539 @@
+        let client = self.inner.weak_client();
+        let cache = self.cache.clone();
+        RpcWithBlock::new_provider(move |block_id| {
+            let req = RequestType::new("eth_getStorageAt", (address, key)).with_block_id(block_id);
+            cache_rpc_call_with_block!(cache, client, req)
+        })
+    }
+}
+
+struct RequestType<Params: RpcParam> {
+    method: &'static str,
+    params: Params,
+    block_id: Option<BlockId>,
+}
+
+impl<Params: RpcParam> RequestType<Params> {
+    const fn new(method: &'static str, params: Params) -> Self {
+        Self { method, params, block_id: None }
+    }
+
+    const fn with_block_id(mut self, block_id: BlockId) -> Self {
+        self.block_id = Some(block_id);
+        self
+    }
+
+    fn params_hash(&self) -> TransportResult<B256> {
+        let hash = serde_json::to_string(&self.params())
+            .map(|p| keccak256(p.as_bytes()))
+            .map_err(RpcError::ser_err)?;
+
+        Ok(hash)
+    }
+
+    const fn method(&self) -> &'static str {
+        self.method
+    }
+
+    fn params(&self) -> Params {
+        self.params.clone()
+    }
+
+    /// Returns true if the BlockId has been set to a tag value such as "latest", "earliest", or
+    /// "pending".
+    const fn has_block_tag(&self) -> bool {
+        if let Some(block_id) = self.block_id {
+            match block_id {
+                BlockId::Hash(_) => return false,
+                BlockId::Number(BlockNumberOrTag::Number(_)) => return false,
+                _ => return true,
+            }
+        }
+        false
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct FsCacheEntry {
+    /// Hash of the request params
+    key: B256,
+    /// Serialized response to the request from which the hash was computed.
+    value: String,
+}
+
+/// Configuration for the cache layer.
+/// For future extensibility of the configurations.
+#[derive(Debug, Clone)]
+pub struct CacheConfig {
+    /// Maximum number of items to store in the cache.
+    pub max_items: usize,
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use crate::ProviderBuilder;
+    use alloy_network::TransactionBuilder;
+    use alloy_node_bindings::Anvil;
+    use alloy_primitives::{Bytes, FixedBytes};
+    use alloy_rpc_types_eth::{BlockId, TransactionRequest};
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_get_accounts() {
+        let cache = CacheLayer::new(100);
+        let anvil = Anvil::new().block_time_f64(0.3).spawn();
+        let provider = ProviderBuilder::default().layer(cache).on_http(anvil.endpoint_url());
+
+        let path = PathBuf::from_str("./rpc-cache-accounts.txt").unwrap();
+        provider.load_cache(path.clone()).unwrap();
+
+        let accounts = provider.get_accounts().await.unwrap();
+        let accounts2 = provider.get_accounts().await.unwrap();
+        assert_eq!(accounts, accounts2);
+
+        provider.save_cache(path).unwrap();
```

Comment:
won't this write a new file and pollute the disk?

Reply:
Yeah changing this to use tmp_dir